### PR TITLE
feat: Add ODP Datafile Parsing and Audience Evaluation

### PR DESCRIPTION
**Pull request by zashraf1985**
Summary
-------

- A new field `integrations` and new audience type `odp.audiences` will be added to the datafile schema for ODP integration.

- Parses new datafile odp values (“publicKey” and “host”) from the “integrations” section. These values are used to send requests to the ODP server.

- Parses datafile with and without the “integrations” section or empty section as well.

-  support a new audience with the `odp.audiences` and `qualified` match.

-  support any (allowed) combination of audiences, including ODP and other existing audiences.

## Test plan
- Added New unit tests
- Existing Tests should pass

---

## Issues
- OASIS-8390

### CODE REVIEW

File: packages/optimizely-sdk/lib/core/audience_evaluator/index.ts [hunk 1]
```diff
@@ -64,7 +66,8 @@ export class AudienceEvaluator {
   evaluate(
     audienceConditions: Array<string | string[]>,
     audiencesById: { [id: string]: Audience },
-    userAttributes: UserAttributes = {}
+    userAttributes: UserAttributes = {},
+    segments: string[] = [],
```
Comment (jaeopt) [inline packages/optimizely-sdk/lib/core/audience_evaluator/index.ts:70]: add 'segments' in doc comments

File: packages/optimizely-sdk/lib/core/audience_evaluator/index.ts [hunk 2]
```diff
@@ -99,14 +102,14 @@ export class AudienceEvaluator {
    * @param  {Condition}            condition          A single condition object to evaluate.
    * @return {boolean|null}                            true if the condition is satisfied, null if a matcher is not found.
    */
-  evaluateConditionWithUserAttributes(userAttributes: UserAttributes, condition: Condition): boolean | null {
+  evaluateConditionWithUserAttributes(userAttributes: UserAttributes, segments: string[], condition: Condition): boolean | null {
```
Comment (jaeopt) [inline packages/optimizely-sdk/lib/core/audience_evaluator/index.ts:105]: update doc comments

**Review by jaeopt — CHANGES_REQUESTED** [review]
Qualified-type support looks good! I'd like to see UserContext passed down to audience evaluators. See my comments and let's chat about it.

File: packages/optimizely-sdk/lib/optimizely_user_context/index.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ export default class OptimizelyUserContext {
   private userId: string;
   private attributes: UserAttributes;
   private forcedDecisionsMap: { [key: string]: { [key: string]: OptimizelyForcedDecision } };
+  private qualifiedSegments: string[] = [];
```
Comment (jaeopt) [inline packages/optimizely-sdk/lib/optimizely_user_context/index.ts:32]: Can we open {get, set} for this prop to public?  They can override it with own segments for debugging or other purposes?

File: packages/optimizely-sdk/lib/core/audience_evaluator/odp_segment_condition_evaluator/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    evaluator = qualifiedEvaluator;
+  } else {
+    evaluator = EVALUATORS_BY_MATCH_TYPE[conditionMatch] || qualifiedEvaluator;
+  }
+
+  return evaluator(condition, user);
+}
+
+function qualifiedEvaluator(condition: Condition, user: OptimizelyUserContext): boolean {
+  //return user.getQualifiedSegments().indexOf(condition.value as string) > -1;
```
Comment (jaeopt) [inline packages/optimizely-sdk/lib/core/audience_evaluator/odp_segment_condition_evaluator/index.ts:63]: remove this comment

File: packages/optimizely-sdk/lib/core/audience_evaluator/index.tests.js [hunk 5]
```diff
@@ -323,12 +332,248 @@ describe('lib/core/audience_evaluator', function() {
...
+                "match": "qualified"
+            }
+          ]
+        };
+        const audiencesById = {
+          0: singleAudience,
+        }
+        const audience = new AudienceEvaluator();
+        
+        it('shoud evaluate to true if segment is found', () => {
```
Comment (jaeopt) [inline packages/optimizely-sdk/lib/core/audience_evaluator/index.tests.js:361]:
```suggestion
        it('should evaluate to true if segment is found', () => {
```
fix same for others below

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good! A couple of small changes suggested.

**Review by jaeopt — APPROVED** [review]
LGTM

**Comment by opti-jnguyen** [comment]
@zashraf1985 - I believe I fixed the import issues, however although the Git Actions shows all tests passing there seems to be 2 failing / 2 pending checks still. What am I missing?

**Pull request merged**